ros2_reduct_agent/recorder.py

```diff
@@ -0,0 +1,376 @@
+            )
+
+            schema = state.schemas_by_topic.get(topic_name)
+            if not schema:
+                self.logger.warn(
+                    f"[{pipeline_name}] No schema registered for topic '{topic_name}'"
+                )
+                continue
+
+            state.writer.write_message(
+                topic=topic_name,
+                schema=schema,
+                message=message,
+                publish_time=publish_time,
+            )
+            state.current_size = state.buffer.tell()
+            split_size = self.pipeline_configs[pipeline_name].split_max_size_bytes
+            if split_size and state.current_size >= split_size:
+                self.upload_pipeline(pipeline_name, state)
+
+    #
+    # Pipeline Management and Upload
+    #
+    def make_timer_callback(self, pipeline_name: str):
+        """Return a callback that uploads the current pipeline state."""
+
+        def _timer_callback():
+            state = self.pipeline_states[pipeline_name]
+            self.upload_pipeline(pipeline_name, state)
+
+        return _timer_callback
+
+    def upload_pipeline(self, pipeline_name: str, state: PipelineState):
+        """Finish current MCAP, upload if it contains data, and reset writer and state."""
+        if not all([state.writer, state.buffer, state.timer]):
+            self.logger.warn(
+                f"[{pipeline_name}] Incomplete state (writer/buffer/timer) — skipping upload."
+            )
+            return
+
+        if state.is_uploading:
+            self.logger.warn(
+                f"[{pipeline_name}] Upload already in progress — skipping upload."
+            )
+            return
+
+        if state.current_size == 0:
+            self.logger.info(
+                f"[{pipeline_name}] No new data since last upload — skipping upload."
+            )
+            state.timer.reset()
+            return
+
+        # Prevent concurrent uploads between timer and max size triggers
+        state.is_uploading = True
+
+        # Filename index either by increment or timestamp in microseconds
+        filename_mode = self.pipeline_configs[pipeline_name].filename_mode
+        file_index = (
+            state.increment
+            if filename_mode == FilenameMode.INCREMENTAL
+            else (state.first_timestamp or self.get_clock().now().nanoseconds) // 1_000
+        )
+
+        # Get the data from the buffer
+        state.writer.finish()
+        state.buffer.seek(0)
+        data = state.buffer.read()
+
+        # Upload and reset the state
+        self.upload_mcap(pipeline_name, data, file_index)
+        self.reset_pipeline_state(pipeline_name, state)
+        state.increment += 1
+
+    def upload_mcap(self, pipeline_name: str, data: bytes, file_index: int):
+        """Upload MCAP to ReductStore."""
+        self.logger.info(f"[{pipeline_name}] MCAP ready. Uploading to ReductStore...")
+
+        async def _upload():
+            await self.bucket.write(
```

Comment:
instead of readying the whole chunk from the buffer, you can write the buffer directly to the bucket:

```python

async def iterator():
   while True:
      chunk = buffer.read(100_000)
      if not chunk:
         break
      yield chunk


await bucket.write(pipeline_name, iterator()...)
```